python-avd/pyavd/api/pool_manager/base_classes.py

```diff
@@ -0,0 +1,282 @@
+# that can be found in the LICENSE file.
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import asdict, dataclass, field
+from typing import TYPE_CHECKING, Generic, TypeVar, cast
+
+from yaml import CSafeDumper, CSafeLoader, dump, load
+
+from pyavd.j2filters import natural_sort
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+    from typing_extensions import Self
+
+    from pyavd._eos_designs.shared_utils import SharedUtils
+
+T_ValueType = TypeVar("T_ValueType", int, str)
+T_PoolKeyType = TypeVar("T_PoolKeyType", bound="PoolKey")
+T_AssignmentKeyType = TypeVar("T_AssignmentKeyType", bound="AssignmentKey")
+
+FILE_HEADER = """\
+# This file is auto-generated by AVD eos_designs.
+# When there is a merge conflict for this file, rebase the branch, accept all current changes for this file, and re-run AVD.
+"""
+
+
+@dataclass(frozen=True)
+class PoolKey:
+    @classmethod
+    def from_dict(cls, data: dict) -> Self:
+        return cls(**data)
+
+
+@dataclass(frozen=True)
+class AssignmentKey:
+    @classmethod
+    def from_dict(cls, data: dict) -> Self:
+        return cls(**data)
+
+
+@dataclass
+class PoolAssignment(Generic[T_AssignmentKeyType, T_ValueType]):
+    key: T_AssignmentKeyType
+    value: T_ValueType
+    accessed: bool = False
+
+    def as_dict(self) -> dict:
+        return {"key": asdict(self.key), "value": self.value}
+
+
+@dataclass
+class Pool(Generic[T_PoolKeyType, T_AssignmentKeyType, T_ValueType]):
+    """One Pool of resources. Currently only supporting a pool of Integers."""
+
+    collection: PoolCollection[T_PoolKeyType, T_AssignmentKeyType, T_ValueType]
+    pool_key: T_PoolKeyType
+    assignments: dict[T_AssignmentKeyType, PoolAssignment[T_AssignmentKeyType, T_ValueType]]
+
+    def _is_value_available(self, value: T_ValueType) -> bool:
+        if self.collection.value_type is int and isinstance(value, int):
+            collection = cast(PoolCollection[T_PoolKeyType, T_AssignmentKeyType, int], self.collection)
+            assignments = cast(dict[T_AssignmentKeyType, PoolAssignment[T_AssignmentKeyType, int]], self.assignments)
+            existing_ids = {assignment.value for assignment in assignments.values()}
+            return not (value in existing_ids or value < collection.min_value or (collection.max_value is not None and value > collection.max_value))
+
+        # currently we only support int pools
+        return False
+
+    def _next_available(self) -> T_ValueType:
+        if self.collection.value_type is int:
+            collection = cast(PoolCollection[T_PoolKeyType, T_AssignmentKeyType, int], self.collection)
+            assignments = cast(dict[T_AssignmentKeyType, PoolAssignment[T_AssignmentKeyType, int]], self.assignments)
+            existing_ids = {assignment.value for assignment in assignments.values()}
+            available_ids = set(range(collection.min_value, len(existing_ids) + 2)) - existing_ids
```

Comment:
I undersrand the logic here though it is not obvious so probably we should add a comment explaining why this is enough to do this.

Also what do you mean by `range` ? here if my `collection.min_value=10` and I have 2 ids say 11 and 12 this yields:

```
>>> existing_ids = {11, 12}
>>> min_value = 10
>>> set(range(min_value, len(existing_ids) + 2)) - existing_ids
set()
>>> range(min_value, len(existing_ids) + 2)
range(10, 4)
```

I think you meant to say a range of length eixting_ids + 2 starting from colleciton.min_value

This does not show up in the people testing the PR because we always start from 1 in the ID pool

Reply:
I think I have improved it a bit now by not creating an extra set from range. It will save on iterations in case of available IDs "in between" other IDs.